### PR TITLE
Add client app

### LINKED ISSUE #56: Re-open client app
#55 
I add client app again in the right branch. 
also made some fixes.

### PR DESCRIPTION
**Pull request by LiorSilberman**
I reopen client PR, now it is in the right place.
This PR contain client module, migrations and tests.

I update conftest.py, I moved my client fixtures to the conftest.py instead of being in test_client.py file
fixes #56

### CODE REVIEW

**Comment by pazMenachem** [comment]
need tests in order to review, contribution file🤷‍♂️

**Comment by MeitarRizner** [comment]
As Paz said - add some tests. Try to use 'Supplier' PR to understand how to do it, the modules are similar.
In addition you need to write a data test (0002_test_data file).

File: client/models.py
Comment (thetaxas2468) [inline client/models.py:40]: Change the min validator to 1,Haifa has 5 chars and not .

**Comment by thetaxas2468** [comment]
link this PR to an issue
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

check out this link

**Comment by RawanAbuleil** [comment]
great job Lior , once the tests are up will be able to comment

**Comment by thetaxas2468** [comment]
> link this PR to an issue https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
> 
> check out this link

Still not linked checkout the link, it should be something like this:
Commit information with an opened issue
<9 options> the issue id

Example:
Information
fixes #55

File: client/models.py
Comment (yftacherzog) [inline client/models.py:83]: You have a race condition here. What happens if the client is deleted between the time you check if it exists and the time it's deleted. It's better to avoid the check and instead wrap the delete in try/except and handle the exception raised if it does not exist.

File: client/tests/test_client.py
Comment (yftacherzog) [inline client/tests/test_client.py:39]: How do you know it wasn't there before?

File: client/tests/test_client.py
Comment (yftacherzog) [inline client/tests/test_client.py:48]: You're asserting with `client0` while you only have `saved_client0` defined on this function. I'm not really sure how it does not raise an exception.

Comment (yftacherzog) [reply client/tests/test_client.py:48]: Try verifying that this line even gets executed (e.g. put `assert False` after it). Might be that there's some sort of magic here, but in most cases it's not that exotic...
In any case, you should use `saved_client0` also for the assertion.

File: client/tests/test_client.py
Comment (yftacherzog) [inline client/tests/test_client.py:58]: Do you actually have any clients there? Maybe verify that `Client.objects.all()` is not empty before.

File: client/tests/test_client.py
Comment (yftacherzog) [inline client/tests/test_client.py:66]: You actually do have an arg here. Generally speaking, you can avoid docstrings for most test functions. The function should be short and with an informative name, and you can use strings and variable names to convey what they are about.

In this example, why is this `user_name` not valid? You can name it in a way that would explain that.

**Comment by LiorSilberman** [comment]
> > link this PR to an issue https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
> > check out this link
> 
> Still not linked checkout the link, it should be something like this: Commit information with an opened issue <9 options> the issue id
> 
> Example: Information fixes #55

Ok thanks just did it

File: client/tests/test_client.py
Comment (LiorSilberman) [reply client/tests/test_client.py:39]: I can add in line 37: assert client0 not in models.Client.objects.all()

so it would prove that client0 not in the table before. 
this is what you mean right?

File: client/tests/test_client.py
Comment (LiorSilberman) [reply client/tests/test_client.py:48]: Thanks, I fixed that

File: client/tests/test_client.py
Comment (yftacherzog) [reply client/tests/test_client.py:39]: yes

File: client/tests/test_client.py
Comment (LiorSilberman) [reply client/tests/test_client.py:66]: In this example the user_name mot valid because we decided that the min length of user_name will be  6 characters.

Comment (yftacherzog) [reply client/tests/test_client.py:66]: You can call it `short` then maybe.

Comment (LiorSilberman) [reply client/tests/test_client.py:66]: Ok I got it

File: client/models.py
Comment (thetaxas2468) [inline client/models.py:82]: delete does not raise does not exist exception,get method does

File: client/tests/test_client.py
Comment (thetaxas2468) [inline client/tests/test_client.py:39]: test deletion of something that does not exist

File: client/tests/test_client.py
Comment (thetaxas2468) [inline client/tests/test_client.py:50]: Test all the validators

**Comment by thetaxas2468** [comment]
Your repo is working in progress because you have things that will be deleted in the future and dont need to be reviewd like User Model and change your repo to ready for review and add a comment that explain what things to be reviewed and what you dont want reviewers to review

File: client/tests/test_client.py
Comment (yftacherzog) [inline client/tests/test_client.py:57]: How do you know which of them failed you? Suppose that tomorrow someone accidentally breaks the validator for the password field, this test will still pass. You need to have each failure on its own test (you can also parameterize this test if you prefer - https://docs.pytest.org/en/6.2.x/parametrize.html).

File: client/migrations/0001_initial.py
Comment (thetaxas2468) [inline client/migrations/0001_initial.py:30]: user type is deleted,please modify the migrations

File: client/migrations/0002_test_data.py
Comment (thetaxas2468) [inline client/migrations/0002_test_data.py:13]: Here you are using user_type as client but user_Type is deleted please update

File: client/migrations/0002_test_data.py
Comment (thetaxas2468) [inline client/migrations/0002_test_data.py:23]: same here

File: client/models.py
Comment (thetaxas2468) [inline client/models.py:30]: user is deleted,Please rebase your commit on upstream/main since User is on the buy_togehter_app import it from there and use it

File: client/tests/test_client.py
Comment (thetaxas2468) [inline client/tests/test_client.py:11]: user_type deleted.

File: client/tests/test_client.py
Comment (thetaxas2468) [inline client/tests/test_client.py:22]: same here

**Comment by yftacherzog** [comment]
> you have 2 commits for your PR can your squash them in one commit

@LiorSilberman a rebase was needed here rather than merge.

**Comment by thetaxas2468** [comment]
LGTM,just solve yftach commment + the validators tests and wait for omer and yftach to approve if there is no things to be changed

**Review by thetaxas2468 — APPROVED** [review]

File: client/models.py
Comment (Omeramsc) [inline client/models.py:20]: Is there a need to wrap the Django builtin method? if not, this method is redundant and can be removed. 
We should not maintain code the Django team already maintain.

File: client/models.py
Comment (Omeramsc) [inline client/models.py:49]: Is there a need to wrap the Django builtin method? if not, this method is redundant and can be removed. 
We should not maintain code the Django team already maintain.

File: client/models.py
Comment (Omeramsc) [inline client/models.py:50]: 1. does delete raise a ValueError?
2. why do we catch the exception if the only thing we're doing with it is raising it again? there is no special handling made (no message to the user, no raising a different error... ) please look at the comments I left for Paz regarding the same subject:  https://github.com/redhat-beyond/BuyTogether/pull/16#discussion_r1049445757

File: client/models.py
Comment (Omeramsc) [inline client/models.py:35]: does delete raise a ValueError?
why do we catch the exception if the only thing we're doing with it is raising it again? there is no special handling made (no message to the user, no raising a different error... ) please look at the comments I left for Paz regarding the same subject: https://github.com/redhat-beyond/BuyTogether/pull/16#discussion_r1049445757

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Good job!  I Left a few comments, but I believe this PR is very close to being merged :)

File: client/models.py
Comment (LiorSilberman) [reply client/models.py:50]: Ok I understood and agree. 
i thought that if I try to delete client that already deleted so his value will be 'None'.. this is why i used ValueError. 

I changed it   👍

File: client/admin.py
Comment (pazMenachem) [inline client/admin.py:7]: user type need to be deleted

File: client/models.py
Comment (pazMenachem) [reply client/models.py:35]: delete doesnt raise an error, full clean does.

File: client/models.py
Comment (pazMenachem) [inline client/models.py:49]:
```suggestion
            Client.object.get(user_name=client.user_name).delete()
```
that will raise an error

Comment (pazMenachem) [reply client/models.py:49]: other than that LGTM

File: client/models.py
Comment (thetaxas2468) [inline client/models.py:39]: In this function you are doing something that maybe raise an error and then you raise it without doing anything so maybe just do:
Client.object.get(user_name=client.user_name).delete()
without try and catch

File: client/models.py
Comment (thetaxas2468) [inline client/models.py:21]: Why this function a static function and not regular object function,if you have the object client already why not just call save on the object self.

File: client/models.py
Comment (thetaxas2468) [inline client/models.py:34]: same as save.

File: client/tests/test_client.py
Comment (yftacherzog) [inline client/tests/test_client.py:43]: Should be in a test case of its own

File: client/tests/test_client.py
Comment (thetaxas2468) [inline client/tests/test_client.py:49]: This test returns true all the time except if you dont have clients,what is the point?

File: client/tests/test_client.py
Comment (Omeramsc) [inline client/tests/test_client.py:52]: we should keep in the context manager only  the lines we expect to raise the exception, in this case line 53. line 52 should be outside the context manager

File: client/tests/test_client.py
Comment (Omeramsc) [inline client/tests/test_client.py:59]: same

File: client/tests/test_client.py
Comment (Omeramsc) [inline client/tests/test_client.py:66]: same

File: client/tests/test_client.py
Comment (Omeramsc) [inline client/tests/test_client.py:73]: same

File: client/tests/test_client.py
Comment (Omeramsc) [inline client/tests/test_client.py:80]: same

**Review by thetaxas2468 — APPROVED** [review]
LGTM!

**Review by Omeramsc — APPROVED** [review]
LGTM
next time, take a look at pytest's [parametrization](https://docs.pytest.org/en/6.2.x/parametrize.html), it can help making tests more readable and easy.

**Pull request merged**
